cranelift-codegen/src/peephole.rs

```diff
@@ -0,0 +1,469 @@
+// would imply that the Value resided in both of those registers.  Function
+// `invariants_ok` checks this.
+//
+// It is also the case that the set of `(offset, size)` segments in the
+// environment must be non-overlapping.  `invariants_ok` doesn't check that,
+// though.
+
+struct AvailEnv {
+    map: Vec<Option<SlotInfo>>,
+    size: usize,
+}
+
+impl AvailEnv {
+    // Create a new one.
+    fn new(size: usize) -> Self {
+        let mut env = AvailEnv {
+            map: Vec::<Option<SlotInfo>>::new(),
+            size: size,
+        };
+        for _ in 0..size {
+            env.map.push(None);
+        }
+        debug_assert!(env.map.len() == size);
+        env
+    }
+
+    // Debug only: checks (some of) the required AvailEnv invariants.
+    fn invariants_ok(&self) -> bool {
+        debug_assert!(self.map.len() == self.size);
+        // This is super lame (quadratic), but it's only used in debug builds.
+        for i in 0..self.size {
+            if let Some(si) = self.map[i] {
+                for j in i + 1..self.size {
+                    if let Some(sj) = self.map[j] {
+                        if si.value == sj.value {
+                            return false;
+                        }
+                    }
+                }
+            }
+        }
+        true
+    }
+
+    // Invalidates the binding associated with `reg`.  Note that by
+    // construction of `AvailEnv`, `reg` can only be associated with one
+    // binding at once.
+    fn invalidate_by_reg(&mut self, reg: RegUnit) {
+        debug_assert!(self.map.len() == self.size);
+        debug_assert!((reg as usize) < self.size);
+        self.map[reg as usize] = None;
+    }
+
+    // Invalidates any binding characterised by `(offs, size)`.  Per
+    // discussion of invariants above, `(offs, size)` can only be associated
+    // with one binding at once.
+    fn invalidate_by_offset(&mut self, offs: i32, size: u32) {
+        debug_assert!(self.map.len() == self.size);
+        for i in 0..self.size {
+            if let Some(si) = &self.map[i] {
+                if any_overlap(&si, offs, size) {
+                    self.map[i] = None;
+                }
+            }
+        }
+    }
+
+    // Invalidates all bindings in the enviroment.
+    fn invalidate_all(&mut self) {
+        debug_assert!(self.map.len() == self.size);
+        for i in 0..self.size {
+            self.map[i] = None;
+        }
+    }
+
+    // Updates the environment to track the effect of an
+    // `InstructionData::RegMove` instruction.  This moves a value from one
+    // register to another, but doesn't create a new SSA name.  Hence we copy
+    // `src`'s entry to `dst` verbatim, without installing a new value in
+    // `dst`'s `Value` field.
+    fn move_reg(&mut self, src: RegUnit, dst: RegUnit) {
+        debug_assert!(self.map.len() == self.size);
+        self.map[dst as usize] = self.map[src as usize];
+        self.map[src as usize] = None;
```

Comment:
That would break the SSA invariant to keep the value alive, right? Makes me wonder if there are missed opportunities because of this, and if the value stored in the map should be either an Option<SlotInfo> or a (new kind of) alias

Reply:
This is quite a subtle point.  `move_reg` is used only for `InstructionData::RegMove`, and the purpose of that instruction is exactly to move a value to a different register without creating a new SSA name.  So it is true that the old register would still have the same value, which we might be able to use later in the EBB to convert a fill into a copy.  But the problem is, doing such a conversion requires having the SSA name associated with the register available.  And in this situation I don't know what it would be -- I assume there would be none.

It might also be the case that regalloc creates `InstructionData::RegMove` exactly because it needs to free up the source register so that it can immediately overwrite it.  In that case, continuing to track the source register as containing a cached fill value is pointless since it'll shortly get overwritten.  I am however merely guessing here, since I don't know the actual purpose of `InstructionData::RegMove`.